File: udp sender/voice_control.py
import socket
import speech_recognition as sr
import struct
import time
from dataclasses import dataclass
from gpiozero import Button  # Import the Button class from gpiozero
from gpiozero.pins.pigpio import PiGPIOFactory

# Define the GPIO pin for the button (adjust as needed)
BUTTON_PIN = 27

@dataclass
class VehicleCommand:
    command: int
    param1: float
    param2: float = 0.0
    target_system: int = 1
    target_component: int = 1
    source_system: int = 1
    source_component: int = 1
    from_external: bool = True

    def pack(self):
        packed_data = struct.pack(
            'IffBBBB?',
            self.command,
            self.param1,
            self.param2,
            self.target_system,
            self.target_component,
            self.source_system,
            self.source_component,
            self.from_external
        )

        return packed_data

class VoiceControlNode:
    def __init__(self, udp_ip, udp_port):
        self.udp_ip = udp_ip
        self.udp_port = udp_port
        
        # Initialize UDP socket
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        
        # Initialize speech recognition
        self.recognizer = sr.Recognizer()
        self.recognizer.energy_threshold = 400
        self.mic = sr.Microphone()
        
        # Command mappings
        self.command_handlers = {
            "alpha": self.arm_drone,
            "thank you": self.disarm_drone,
            "off board": self.switch_to_offboard_mode,
            "gripper open": self.open_gripper,
            "gripper close": self.close_gripper,
            "go up": self.throttle_up,
            "go down": self.throttle_down,
            "turn left": self.yaw_left,
            "turn right": self.yaw_right,
            "move forward": self.move_forward,
            "stop": self.stop_movement   # New stop command via voice
        }
        
        # Set up button using gpiozero
        factory = PiGPIOFactory()
        self.button = Button(BUTTON_PIN, pin_factory=factory)
        self.button.when_pressed = self.gpio_stop_callback  # Trigger callback on button press

    def gpio_stop_callback(self):
        # This callback is triggered when the button is pressed.
        print("GPIO button pressed, triggering stop command.")
        self.stop_movement()

    def send_command(self, cmd: VehicleCommand):
        try:
            sent = self.sock.sendto(cmd.pack(), (self.udp_ip, self.udp_port))
            print(f"Command sent via UDP: {cmd.command}")
        except Exception as e:
            print(f"Failed to send UDP message: {e}")

    def arm_drone(self):
        cmd = VehicleCommand(
            command=400,  # MAV_CMD_COMPONENT_ARM_DISARM
            param1=1.0    # 1 to arm
        )
        self.send_command(cmd)
        print("Arm command sent")
        send_udp("ARMED")

    def disarm_drone(self):
        cmd = VehicleCommand(
            command=400,  # MAV_CMD_COMPONENT_ARM_DISARM
            param1=0.0    # 0 to disarm
        )
        self.send_command(cmd)
        print("Disarm command sent")
        send_udp("DISARM")

    def switch_to_offboard_mode(self):
        cmd = VehicleCommand(
            command=176,  # VEHICLE_CMD_DO_SET_MODE
            param1=1.0,
            param2=6.0    # OFFBOARD mode
        )
        self.send_command(cmd)
        print("Offboard mode command sent")
        send_udp("OFFBOARD")

    def open_gripper(self):
        cmd = VehicleCommand(
            command=187,  # Gripper command
            param1=1.0,
            param2=1.0
        )
        self.send_command(cmd)
        print("Gripper open command sent")
        send_udp("GRIPPER OPEN")

    def close_gripper(self):
        cmd = VehicleCommand(
            command=187,  # Gripper command
            param1=-1.0,
            param2=-1.0
        )
        self.send_command(cmd)
        print("Gripper close command sent")
        send_udp("GRIPPER CLOSE")

    def throttle_up(self):
        cmd = VehicleCommand(
            command=178,  # Custom throttle command
            param1=2,
            param2=-1
        )
        self.send_command(cmd)
        print("Throttle up command sent")
        send_udp("UP")
        
    def throttle_down(self):
        cmd = VehicleCommand(
            command=178,  # Custom throttle command
            param1=2,
            param2=1
        )
        self.send_command(cmd)
        print("Throttle down command sent")
        send_udp("DOWN")

    def yaw_left(self):
        cmd = VehicleCommand(
            command=179,  # Custom yaw command
            param1=-1.0
        )
        self.send_command(cmd)
        print("Yaw left command sent")
        send_udp("LEFT")

    def yaw_right(self):
        cmd = VehicleCommand(
            command=179,  # Custom yaw command
            param1=1.0
        )
        self.send_command(cmd)
        print("Yaw right command sent")
        send_udp("RIGHT")

    def move_forward(self):
        cmd = VehicleCommand(
            command=180,  # Custom forward movement command
            param1=10
        )
        self.send_command(cmd)
        print("Move forward command sent")
        send_udp("FORWARD")
        
    def stop_movement(self):
        # New stop command: command 181 stops forward movement (holds current position)
        cmd = VehicleCommand(
            command=181,  # Custom stop command
            param1=0.0
        )
        self.send_command(cmd)
        print("Stop command sent")
        send_udp("STOP")

    def run(self):
        print("Voice control running. Press Ctrl+C to exit.")
        # Adjust for ambient noise and listen for commands
        with self.mic as source:
            print("Adjusting for ambient noise...")
            self.recognizer.adjust_for_ambient_noise(source, duration=1)
            
            try:
                while True:
                    print("Listening...")
                    audio = self.recognizer.listen(source)
                    
                    try:
                        # Recognize speech using Google Web Speech API
                        text = self.recognizer.recognize_google(audio, language="en-EN")
                        print(f"Recognized: {text}")
                        
                        # Check for commands in the recognized text
                        text_lower = text.lower()
                        for command, handler in self.command_handlers.items():
                            if command in text_lower:
                                print(f"Command detected: {command}")
                                handler()
                                
                    except sr.UnknownValueError:
                        print("Could not understand audio")
                    except sr.RequestError as e:
                        print(f"Could not request results from Google Speech Recognition service; {e}")
            except KeyboardInterrupt:
                print("\nStopping voice control...")

    def cleanup(self):
        # Clean up resources if needed
        GPIO.cleanup()

def send_udp(message):
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    message_bytes = message.encode('utf-8')
    ip_address = "127.0.0.1"
    port = 7880
    bytes_sent = sock.sendto(message_bytes, (ip_address, port))
    return bytes_sent
    
def main():
    import sys
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <udp_ip> <udp_port>")
        sys.exit(1)
        
    try:
        controller = VoiceControlNode(sys.argv[1], int(sys.argv[2]))
        controller.run()
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: udp sender/voice_control.cpp
#include <iostream>
#include <string>
#include <thread>
#include <regex>
#include <algorithm>
#include <cstring>
#include <unistd.h>
#include <arpa/inet.h>
#include <wiringPi.h>

#define BUTTON_PIN 27  // GPIO Pin for physical stop button

struct VehicleCommand {
    uint32_t command;
    float param1;
    float param2;
    uint8_t target_system;
    uint8_t target_component;
    uint8_t source_system;
    uint8_t source_component;
    bool from_external;

    void pack(uint8_t* buffer) {
        std::memcpy(buffer, this, sizeof(VehicleCommand));
    }
};

class VoiceControl {
public:
    VoiceControl(const std::string& udp_ip, int udp_port) : udp_ip_(udp_ip), udp_port_(udp_port) {
        setup_udp_socket();
        setup_gpio();
        monitor_whisper_output();
    }

    ~VoiceControl() {
        close(udp_socket_);
    }

private:
    std::string udp_ip_;
    int udp_port_;
    int udp_socket_;
    std::thread whisper_monitor_thread_;
    std::regex heard_pattern_ = std::regex("Heard '([^']+)'");  // Regex to extract recognized text

    void setup_udp_socket() {
        udp_socket_ = socket(AF_INET, SOCK_DGRAM, 0);
        if (udp_socket_ < 0) {
            std::cerr << "Failed to create UDP socket" << std::endl;
            exit(1);
        }
    }

    void send_udp(const VehicleCommand& cmd) {
        struct sockaddr_in server_addr{};
        server_addr.sin_family = AF_INET;
        server_addr.sin_port = htons(udp_port_);
        inet_pton(AF_INET, udp_ip_.c_str(), &server_addr.sin_addr);

        uint8_t buffer[sizeof(VehicleCommand)];
        cmd.pack(buffer);

        ssize_t sent = sendto(udp_socket_, buffer, sizeof(buffer), 0, 
                              (struct sockaddr*)&server_addr, sizeof(server_addr));
        if (sent < 0) {
            std::cerr << "Failed to send UDP packet" << std::endl;
        } else {
            std::cout << "Sent command: " << cmd.command << std::endl;
        }
    }

    void setup_gpio() {
        wiringPiSetupGpio();
        pinMode(BUTTON_PIN, INPUT);
        pullUpDnControl(BUTTON_PIN, PUD_UP);
        std::thread([this]() {
            while (true) {
                if (digitalRead(BUTTON_PIN) == LOW) {  // Button pressed
                    std::cout << "GPIO button pressed! Sending stop command." << std::endl;
                    stop_movement();
                    std::this_thread::sleep_for(std::chrono::seconds(1));
                }
            }
        }).detach();
    }

    void monitor_whisper_output() {
        std::cout << "Starting Whisper.cpp output monitor" << std::endl;
        std::string line;
        std::cin.sync_with_stdio(false);

        while (true) {
            if (std::getline(std::cin, line)) {
                std::cout << "Read line: " << line << std::endl;
                std::smatch matches;
                if (std::regex_search(line, matches, heard_pattern_) && matches.size() > 1) {
                    std::string transcription = matches[1].str();
                    std::cout << "Extracted transcription: " << transcription << std::endl;
                    process_whisper_output(transcription);
                }
            } else {
                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
            }
        }
    }

    void process_whisper_output(const std::string& transcription) {
        std::string lower_text = transcription;
        std::transform(lower_text.begin(), lower_text.end(), lower_text.begin(), ::tolower);
        lower_text.erase(std::remove(lower_text.begin(), lower_text.end(), '.'), lower_text.end());

        if (lower_text.find("alpha") != std::string::npos) {
            arm_drone();
        } else if (lower_text.find("drone alpha") != std::string::npos) {
            disarm_drone();
        } else if (lower_text.find("off board") != std::string::npos) {
            switch_to_offboard_mode();
        } else if (lower_text.find("gripper open") != std::string::npos) {
            open_gripper();
        } else if (lower_text.find("gripper close") != std::string::npos) {
            close_gripper();
        } else if (lower_text.find("go up") != std::string::npos) {
            throttle_up();
        } else if (lower_text.find("go down") != std::string::npos) {
            throttle_down();
        } else if (lower_text.find("turn left") != std::string::npos) {
            yaw_left();
        } else if (lower_text.find("turn right") != std::string::npos) {
            yaw_right();
        } else if (lower_text.find("move forward") != std::string::npos) {
            move_forward();
        } else if (lower_text.find("stop") != std::string::npos) {
            stop_movement();
        }
    }

    void arm_drone() {
        VehicleCommand cmd = {400, 1.0};
        send_udp(cmd);
    }

    void disarm_drone() {
        VehicleCommand cmd = {400, 0.0};
        send_udp(cmd);
    }

    void switch_to_offboard_mode() {
        VehicleCommand cmd = {176, 1.0, 6.0};
        send_udp(cmd);
    }

    void open_gripper() {
        VehicleCommand cmd = {187, 1.0, 1.0};
        send_udp(cmd);
    }

    void close_gripper() {
        VehicleCommand cmd = {187, -1.0, -1.0};
        send_udp(cmd);
    }

    void throttle_up() {
        VehicleCommand cmd = {178, 2.0, -1.0};
        send_udp(cmd);
    }

    void throttle_down() {
        VehicleCommand cmd = {178, 2.0, 1.0};
        send_udp(cmd);
    }

    void yaw_left() {
        VehicleCommand cmd = {179, -1.0};
        send_udp(cmd);
    }

    void yaw_right() {
        VehicleCommand cmd = {179, 1.0};
        send_udp(cmd);
    }

    void move_forward() {
        VehicleCommand cmd = {180, 10.0};
        send_udp(cmd);
    }

    void stop_movement() {
        VehicleCommand cmd = {181, 0.0};
        send_udp(cmd);
    }
};

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <udp_ip> <udp_port>" << std::endl;
        return 1;
    }

    try {
        std::string udp_ip = argv[1];
        int udp_port = std::stoi(argv[2]);
        VoiceControl voiceControl(udp_ip, udp_port);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
